File: algorithms/trees/tarjans-offline-lca/java/TarjansOfflineLCA.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TarjansOfflineLCA {
    public static int[] offlineLca(int n, int[][] edges, int[][] queries) {
        if (n <= 0) {
            return new int[0];
        }

        List<List<Integer>> adjacency = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adjacency.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adjacency.get(edge[0]).add(edge[1]);
            adjacency.get(edge[1]).add(edge[0]);
        }

        int[] parent = new int[n];
        int[] depth = new int[n];
        Arrays.fill(parent, -1);
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        queue.add(0);
        parent[0] = 0;

        while (!queue.isEmpty()) {
            int node = queue.removeFirst();
            for (int next : adjacency.get(node)) {
                if (parent[next] != -1) {
                    continue;
                }
                parent[next] = node;
                depth[next] = depth[node] + 1;
                queue.addLast(next);
            }
        }

        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            result[i] = lca(queries[i][0], queries[i][1], parent, depth);
        }
        return result;
    }

    private static int lca(int a, int b, int[] parent, int[] depth) {
        int x = a;
        int y = b;
        while (depth[x] > depth[y]) {
            x = parent[x];
        }
        while (depth[y] > depth[x]) {
            y = parent[y];
        }
        while (x != y) {
            x = parent[x];
            y = parent[y];
        }
        return x;
    }
}

File: algorithms/trees/tarjans-offline-lca/kotlin/TarjansOfflineLca.kt
fun offlineLca(n: Int, edges: Array<IntArray>, queries: Array<IntArray>): IntArray {
    val adjacency = Array(n) { mutableListOf<Int>() }
    for (edge in edges) {
        if (edge.size >= 2) {
            val u = edge[0]
            val v = edge[1]
            adjacency[u].add(v)
            adjacency[v].add(u)
        }
    }

    val parent = IntArray(n) { -1 }
    val depth = IntArray(n)
    val queue = ArrayDeque<Int>()
    queue.addLast(0)
    parent[0] = 0

    while (queue.isNotEmpty()) {
        val node = queue.removeFirst()
        for (next in adjacency[node]) {
            if (parent[next] != -1) {
                continue
            }
            parent[next] = node
            depth[next] = depth[node] + 1
            queue.addLast(next)
        }
    }

    fun lca(a: Int, b: Int): Int {
        var u = a
        var v = b
        while (depth[u] > depth[v]) {
            u = parent[u]
        }
        while (depth[v] > depth[u]) {
            v = parent[v]
        }
        while (u != v) {
            u = parent[u]
            v = parent[v]
        }
        return u
    }

    return IntArray(queries.size) { index ->
        val query = queries[index]
        lca(query[0], query[1])
    }
}
